keystorage/yubihsm2/src/main/java/tech/pegasys/signers/yubihsm2/YubiHsm2.java

```diff
@@ -0,0 +1,120 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.signers.yubihsm2;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.commons.io.IOUtils;
+import org.zeroturnaround.exec.ProcessExecutor;
+import org.zeroturnaround.exec.ProcessResult;
+
+/**
+ * Access secrets/opaque data from Yubi HSM 2. https://developers.yubico.com/YubiHSM2/. It is
+ * assumed that yubi hsm sdk is installed on the system which contains yubihsm-shell. This class
+ * uses yubihsm-shell, which is a thin wrapper over libyubihsm.
+ */
+public class YubiHsm2 {
+  private final List<String> yubiHsmShellPathArgs;
+  private final String connectorUrl;
+  private final short authKey;
+  private final String password;
+  private final Optional<String> caCert;
+  private final Optional<String> proxy;
+  private final Optional<Map<String, String>> additionalEnvVars;
+
+  /**
+   * Parameters that will be passed to yubihsm-shell process to fetch opaque data
+   *
+   * @param yubiHsmShellPathArgs Path to yubihsm-shell binary and any additional arguments to pass.
+   * @param additionalEnvVars Optional environment variables map that can be passed to yubihsm-shell
+   * @param connectorUrl YubiHSM connector URL. http://127.0.01:12345 or yhusb://serial=123456
+   * @param authKeyObjId The auth key object id
+   * @param password Password
+   * @param caCert Optional CA Cert if connector is running in https mode
+   * @param proxy Optional proxy url if a proxy should be used to access connector url
+   */
+  public YubiHsm2(
+      final List<String> yubiHsmShellPathArgs,
+      final Optional<Map<String, String>> additionalEnvVars,
+      final String connectorUrl,
+      final short authKeyObjId,
+      final String password,
+      Optional<String> caCert,
+      Optional<String> proxy) {
+    this.yubiHsmShellPathArgs = yubiHsmShellPathArgs;
+    this.additionalEnvVars = additionalEnvVars;
+    this.connectorUrl = connectorUrl;
+    this.authKey = authKeyObjId;
+    this.password = password;
+    this.caCert = caCert;
+    this.proxy = proxy;
+  }
+
+  /**
+   * Fetch opaque data from yubi hsm 2
+   *
+   * @param opaqueObjId The object id of opque
+   * @return The data stored against opaque object id (in hex format)
```

Comment:
are the secrets require to be strings? l.e _could_ the yubi store _anything_ - i.e what if its not a hex string, what if its my credit card number? 
I guess I mean - we should simple return the string value stored in the secret, unmodified (rather than stating it will be hex encoded).

Reply:
yubi takes opaque data in following formats:
"base64", "binary", "PEM", "password", "hex", "ASCII". I think we can use ASCII instead of hex that I am currently using (as hex maps directly with BLS/SECP in hex encoding).

Reply:
We have to tell yubihsm which format we want the data to be stored and returned. The default for input and output is binary (i.e. reading from a file).

Reply:
Refactored slightly so that we can pass optional output format. Defaults to `default`.